### ISSUE TITLE
Proper way to clip gradient norm

### ISSUE THREAD

**Issue by pclucas14** [facebookresearch/vissl#213]
## ❓ How to do something using VISSL

Using the Adam optimizer, I want to know what is the best way to clip the gradient norm  after calling backward (using `torch.nn.utils.clip_grad_norm_`). I looked into the `standard_train_step.py` file, but there does not seem to be any gradient clipping in place. All I could find were LARC specific clipping arguments in the config files. 

For my personal use I had to design a custom training step, so I integrated a line with gradient clipping. Under this approach, what is the best/recommended way to pass a grad_clip argument from the `.yaml` file such that it is accessible through the `task` object passed into the `train_step` function ? I am asking because there seems to *already* be a default argument for that. When debugging and printing `vars(args).keys()` there is a `clip_grad_norm` key, however I haven't seen it in the config files. Is it possible to set it ? if so, how ?

Thanks for the amazing support so far, 
Lucas

**Comment by growlix** [comment]
Hi Lucas!

You can see how to control gradient clipping [here](https://github.com/facebookresearch/vissl/blob/2d67e1986cec4c2ac22b517c2490e0c362704fbb/vissl/config/defaults.yaml#L428).

To enable gradient clipping, simply set `USE_GRAD_CLIP: True`, and your desired `NORM_TYPE` and `MAX_NORM`.

It's just a few a lines in the config, so it's easy to miss. Sorry you had to go to so much work implementing it yourself!

**Comment by pclucas14** [comment]
thanks for the quick answer! Could you also point me to where the clipping actually happens ? I'm trying to get a better understanding of how VISSL works

**Comment by pclucas14** [comment]
Wow I actually searched for the default file :p I looked only in `vissl/configs` instead of `vissl/vissl/configs`. I'll know for next time!

**Comment by growlix** [comment]
The [gradient clipping implementation](https://github.com/facebookresearch/vissl/blob/2d67e1986cec4c2ac22b517c2490e0c362704fbb/vissl/hooks/grad_clip_hooks.py) is pretty simple. Its basically just a backwards hook that calls [torch.nn.utils.clip_grad_norm_](https://pytorch.org/docs/stable/generated/torch.nn.utils.clip_grad_norm_.html). If you're using `trainer.train_steps.standard_train_step` (which you should by default, and iirc, because there currently aren't any other types of `train_step`), then backwards hooks are called [here](https://github.com/facebookresearch/vissl/blob/2d67e1986cec4c2ac22b517c2490e0c362704fbb/vissl/trainer/train_steps/standard_train_step.py#L205).

**Comment by pclucas14** [comment]
ahh ok I see, that make sense. Thanks again
